python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,206 @@
+
+    if pd.api.types.is_dtype_equal(ltype, rtype):
+        return lcol, rcol
+
+    if (np.issubdtype(ltype, np.number)) and (np.issubdtype(rtype, np.number)):
+        common_type = (
+            max(ltype, rtype)
+            if ltype.kind == rtype.kind
+            else np.find_common_type([], (ltype, rtype))
+        )
+
+    elif np.issubdtype(ltype, np.datetime64) and np.issubdtype(
+        rtype, np.datetime64
+    ):
+        common_type = max(ltype, rtype)
+
+    if how == "left":
+        if rcol.fillna(0).can_cast_safely(ltype):
+            return lcol, rcol.astype(ltype)
+        else:
+            warnings.warn(
+                f"Can't safely cast column from {rtype} to {ltype}, "
+                "upcasting to {common_type}."
+            )
+
+    return lcol.astype(common_type), rcol.astype(common_type)
+
+
+def _match_categorical_dtypes(
+    lcol: ColumnBase, rcol: ColumnBase, how: str
+) -> Tuple[ColumnBase, ColumnBase]:
+    # cast the keys lcol and rcol to a common dtype
+    # when at least one of them is a categorical type
+    ltype, rtype = lcol.dtype, rcol.dtype
+
+    if isinstance(lcol, cudf.core.column.CategoricalColumn) and isinstance(
+        rcol, cudf.core.column.CategoricalColumn
+    ):
+        # if both are categoricals, logic is complicated:
+        return _match_categorical_dtypes_both(lcol, rcol, how)
+
+    if isinstance(ltype, CategoricalDtype):
+        if how in {"left", "leftsemi", "leftanti"}:
+            return lcol, rcol.astype(ltype)
+        common_type = ltype.categories.dtype
+    elif isinstance(rtype, CategoricalDtype):
+        common_type = rtype.categories.dtype
+    return lcol.astype(common_type), rcol.astype(common_type)
+
+
+def _match_categorical_dtypes_both(
+    lcol: CategoricalColumn, rcol: CategoricalColumn, how: str
+) -> Tuple[ColumnBase, ColumnBase]:
+    # The commontype depends on both `how` and the specifics of the
+    # categorical variables to be merged.
+
+    ltype, rtype = lcol.dtype, rcol.dtype
+
+    # when both are ordered and both have the same categories,
+    # no casting required:
+    if ltype == rtype:
+        return lcol, rcol
+
+    # Merging categorical variables when only one side is ordered is
+    # ambiguous and not allowed.
+    if ltype.ordered != rtype.ordered:
+        raise TypeError(
+            "Merging on categorical variables with mismatched"
+            " ordering is ambiguous"
+        )
+
+    if ltype.ordered and rtype.ordered:
+        # if we get to here, categories must be what causes the
+        # dtype equality check to fail. And we can never merge
+        # two ordered categoricals with different categories
+        raise TypeError(
+            f"{how} merge between categoricals with "
+            "different categories is only valid when "
+            "neither side is ordered"
+        )
+
+    # the following should now always hold
+    assert not ltype.ordered and not rtype.ordered
```

Comment:
Should we explicitly check this and raise instead of asserting?

Reply:
I'll try to make the case for this assertion: it's a confidence check. If this doesn't hold, we did something wrong, not the user. We raise more meaningful error messages about orderedness above, but if we got to here and this condition doesn't hold, there's nothing meaningful we tell the user other than maybe "something went wrong, please raise an issue?". We could "safely" delete this assertion, but it will be helpful for us (the developers) when making changes to the code in the future.